[android] Fix scopedContext unavailable in Sweet API modules

aleqsio opened the pull request:
Fixes: https://github.com/expo/expo/issues/18760

# Why

Trying to access image from ImagePicker via FileSystem results in Location 'file:///…' isn't readable. error.

Appears to be an issue with the scoping of cache files in Expo Go only).

File name looks like this in SDK 45: `file:///data/user/0/host.exp.exponent/cache/ExperienceData/%2540keith-kurak%252Fandroid-imagepicker-uri-unreadable/ImagePicker/4151a827-6c11-4d2d-ab5e-2a4015e8a29c.jpg`

Looks like this in 46: `file:///data/user/0/host.exp.exponent/cache/ImagePicker/48faff81-88bd-4578-86c0-b885f1364c3a.jpeg`

This URI is inaccessible to expo-file-system due to logic defined in `ScopedContext.kt `

This logic prevents activites to access each other's cache as an internal permissions mechanism for virtual, separate experiences in one android app in Expo go.

---
The same `context.cacheDir` is used in 3 modules written in the new API:
- Clipboard
- ImageView
- ImagePicker

However, calendar only uses it as internal cache and cannot save to scoped cache due to the dir being allowlisted in a static xml file. As such it was not changed.

# How

With @lukmccall we attempted to come up with a fix by first passing scoped context to the new style modules. Hard to do, as we would need to make `ScopedContext` inherit from `ReactApplicationContext`, which would be doable if we moved `ScopedContext` only to versioned code (large refactor).

We could make the modules use the ScopedContext directly, but this breaks module contract (they should not rely on the fact they they are scoped (from what I understand).

Instead, we add a new filesystem internal module `ScopedDirectoriesModule` that is instantiated by the old API and is providing scoped paths to modules with new API.

```
  private val scopedDirectories: ScopedDirectories
    get() = appContext.scopedDirectories?.scopedDirectories ?: throw ModuleNotFoundException("ScopedDirectories")
```

This is a stopgap fix until we can come up with a clean way to handle scope in new style modules in the future.


# Test Plan

We can't add tests to BareExpo as it does not contain the Scoping logic. @lukmccall any ideas?
[x] Test manually if the original issue is fixed in Expo Go android for imagepicker. (tested using expo-go built locally and a custom project, simillar to https://snack.expo.dev/@aleqsio/expo-directories-internal-module-test) On main it fails to find experience name in cache dir, on this branch it works as expected.
[x] Test manually if the original issue is fixed in Expo Go android for all 2 modules.
[] Test manually if IOS is unaffected.
[x] Test manually if android standalone/bare is unaffected for all 2 modules.

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md). - should not require doc changes
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

aleqsio committed:
Add scoped directories module to fix scopedContext unavailable in Sweet API modules

aleqsio committed:
Remove unused property

lukmccall commented on packages/expo-modules-core/android/src/main/java/expo/modules/interfaces/filesystem/ScopedDirectories.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Objects;
+
+public final class ScopedDirectories {
+    private final File cacheDir;
+    private final File filesDir;
+
+    public ScopedDirectories(File cacheDir, File filesDir) {
+        this.cacheDir = cacheDir;
+        this.filesDir = filesDir;
+    }
+
+    public File getCacheDir() {
+        return cacheDir;
+    }
+
+    public File getFilesDir() {
+        return filesDir;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == this) return true;
+        if (obj == null || obj.getClass() != this.getClass()) return false;
+        var that = (ScopedDirectories) obj;
+        return Objects.equals(this.cacheDir, that.cacheDir) &&
+                Objects.equals(this.filesDir, that.filesDir);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(cacheDir, filesDir);
+    }
+
+    @Override
+    public String toString() {
+        return "ScopedDirectories[" +
+                "cacheDir=" + cacheDir + ", " +
+                "filesDir=" + filesDir + ']';
+    }
+}
```
```suggestion
data class ScopedDirectories(val cacheDir: File, val filesDir: File)
```
By using Kotlin here, we get the same functionality in just one line of code.

aleqsio replied:
Sorry, still forgetting that I can mix and match kotlin and java and just followed the lang used in the template module I was using. This is actually codegen from Java record classes that is not yet in this Java version (i think), but kotlin data classes work great!

lukmccall reviewed (approved):
Everything looks great. I think that we should avoid using the "scoped" word here. For outside developers, it doesn't make sense. That word is only used internally. So the public interface `ScopedDirectoriesModuleInterface` should be called `DirectoriesModuleInterface` or even `DirectoriesInterface`. Other occurrences of the "scoped" word also have to be removed.

lukmccall reviewed (changes requested):
Sorry, I didn't mean to approve that

aleqsio committed:
Rename ScopedDirectories to directories

lukmccall commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt:
```diff
@@ -132,6 +133,13 @@ class AppContext(
+  val directories: DirectoriesModuleInterface?
+    get() = legacyModule()
+
+
```
```suggestion
```

lukmccall commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/MediaHandler.kt:
```diff
@@ -65,8 +69,9 @@ internal class MediaHandler(
     sourceUri: Uri,
-  ): ImagePickerResponse.Single.Video {
-    val outputFile = createOutputFile(context.cacheDir, ".mp4")
+  ): ImagePickerResponse.Single.Video
+  {
```
```suggestion
  ): ImagePickerResponse.Single.Video {
```

aleqsio committed:
Remove change from clipboard module

As clipboard has a fixed allowed path in xml clipboard_provider_paths, it cannot be scoped. No need to scope it, as it's only a temporary directory.

aleqsio committed:
Fix lint

aleqsio committed:
Run spotless apply

tsapeta commented on packages/expo-modules-core/CHANGELOG.md:
```diff
@@ -33,6 +33,7 @@
 - Add the `RegisterActivityContracts` component to register all of activity result contracts on Android. ([#19180](https://github.com/expo/expo/pull/19180) by [@lukmccall](https://github.com/lukmccall))
 - Improves JSI/JNI type conversion to support complex function arguments on Android. ([#19120](https://github.com/expo/expo/pull/19120) & [#19094](https://github.com/expo/expo/pull/19094) by [@lukmccall](https://github.com/lukmccall))
 - Using JSI instead of the bridge to call native methods also on legacy modules on iOS. ([#19209](https://github.com/expo/expo/pull/19209) by [@lukmccall](https://github.com/lukmccall))
+- Added a `DirectoriesModule` to android as a temporary solution to fix cache directories being incorrect in new Sweet API modules. 
```
```suggestion
- Added a `DirectoriesModule` to Android as a temporary solution to fix cache directories being incorrect in new Sweet API modules. ([#19205](https://github.com/expo/expo/pull/19205) by [@aleqsio](https://github.com/aleqsio))
```

aleqsio committed:
Apply suggestions from code review

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

tsapeta commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt:
```diff
@@ -151,6 +152,12 @@ class AppContext(
+  /**
+   * Provides access to the scoped directories from the legacy module registry.
+   */
+  val directories: DirectoriesModuleInterface?
```
Since the directories module has only one thing, maybe it's better to get these directories right away from here instead of then referring to it as `directories.directories` 😅 
I would also think about changing the name to "app directories"; "directories" is too vague.

tsapeta replied:
Yeah, `appContext.cacheDirectory` and `appContext.documentDirectory` would be great 😉 (for consistency let's use the same names as FileSystem module)

tsapeta commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/DirectoriesModule.kt:
```diff
@@ -0,0 +1,25 @@
+ */
+
+// The class needs to be 'open', because it's inherited in expoview
+open class DirectoriesModule(private val context: Context) : DirectoriesModuleInterface, InternalModule {
```
I'd also change the module name to `AppDirectoriesModule`

tsapeta commented on packages/expo-file-system/CHANGELOG.md:
```diff
@@ -8,6 +8,8 @@
 
 ### 🎉 New features
 
+- Added `DirectoriesModule` to expo-file-system on android as a temporary solution to fix cache directories being incorrect in new Sweet API modules. ([#19205](https://github.com/expo/expo/pull/19205) by [@aleqsio](https://github.com/aleqsio))
```
```suggestion
- Added `DirectoriesModule` to expo-file-system on Android as a temporary solution to fix cache directories being incorrect in new Sweet API modules. ([#19205](https://github.com/expo/expo/pull/19205) by [@aleqsio](https://github.com/aleqsio))
```

aleqsio committed:
Make module private, add property accessors

aleqsio committed:
Add comments

aleqsio commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt:
```diff
@@ -151,6 +154,24 @@ class AppContext(
+  /**
+   * Provides access to the scoped directories from the legacy module registry.
+   */
+  private val appDirectories: AppDirectoriesModuleInterface?
```
Not sure if this should be private (as this currently can be used for nothing else then the properties surfaced below)

lukmccall replied:
Let's leave it private for now ;)

aleqsio committed:
Run spotless check

aleqsio committed:
Update changelog

aleqsio committed:
Change wording

aleqsio committed:
[docs] Fix typo in changelog

aleqsio committed:
Add missing changelogs
